File: src/main/java/ru/ioleksiv/telegram/bot/api/annotations/filter/telegram/StickerFilter.java
package ru.ioleksiv.telegram.bot.api.annotations.filter.telegram;

import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.IntegerFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.StringFilter;
import ru.ioleksiv.telegram.bot.api.model.annotation.AnnotationState;
import ru.ioleksiv.telegram.bot.api.model.annotation.CustomValidator;
import ru.ioleksiv.telegram.bot.api.model.annotation.stub.StubStickerValidator;
import ru.ioleksiv.telegram.bot.api.model.objects.std.sticker.Sticker;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Documented
public @interface StickerFilter {

    AnnotationState status() default AnnotationState.ON;

    Class<? extends CustomValidator<Sticker>>[] validator() default StubStickerValidator.class;

    MaskPositionFilter maskPosition() default @MaskPositionFilter(status = AnnotationState.OFF);

    IntegerFilter width() default @IntegerFilter(status = AnnotationState.OFF);

    IntegerFilter height() default @IntegerFilter(status = AnnotationState.OFF);

    PhotoFilter thumb() default @PhotoFilter(status = AnnotationState.OFF);

    StringFilter emoji() default @StringFilter(status = AnnotationState.OFF);

    StringFilter setName() default @StringFilter(status = AnnotationState.OFF);

    IntegerFilter fileSize() default @IntegerFilter(status = AnnotationState.OFF);

}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/annotations/filter/composite/StickerFilter.kt
package io.github.oleksivio.tl.kbot.core.annotations.filter.composite

import io.github.oleksivio.tl.kbot.core.annotations.filter.primitive.IntegerFilter
import io.github.oleksivio.tl.kbot.core.annotations.filter.primitive.StringFilter
import io.github.oleksivio.tl.kbot.core.model.annotation.AnnotationState

@Retention
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@MustBeDocumented
annotation class StickerFilter(
    val status: AnnotationState = AnnotationState.ON,
    val validator: Array<String> = [],
    val maskPosition: MaskPositionFilter = MaskPositionFilter(
        status = AnnotationState.OFF
    ),
    val width: IntegerFilter = IntegerFilter(
        status = AnnotationState.OFF
    ),
    val height: IntegerFilter = IntegerFilter(
        status = AnnotationState.OFF
    ),
    val thumb: PhotoFilter = PhotoFilter(
        status = AnnotationState.OFF
    ),
    val emoji: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val setName: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val fileSize: IntegerFilter = IntegerFilter(
        status = AnnotationState.OFF
    )
)

